src/Simulation/Simulators.Tests/RuntimeMetadataTests.cs

```diff
@@ -0,0 +1,671 @@
+            Qubit target = new FreeQubit(1);
+            var op = new QuantumSimulator().Get<Intrinsic.X>().Controlled;
+            var args = op.__dataIn((controls, target));
+            var expected = new RuntimeMetadata()
+            {
+                Label = "X",
+                FormattedNonQubitArgs = "",
+                IsAdjoint = false,
+                IsControlled = true,
+                IsMeasurement = false,
+                IsComposite = false,
+                Children = null,
+                Controls = controls,
+                Targets = new List<Qubit>() { target },
+            };
+
+            Assert.True(op.GetRuntimeMetadata(args) == expected);
+        }
+
+        [Fact]
+        public void ControlledCNOT()
+        {
+            IQArray<Qubit> controls = new QArray<Qubit>(new[] { new FreeQubit(0) });
+            Qubit control = new FreeQubit(1);
+            Qubit target = new FreeQubit(2);
+            var op = new QuantumSimulator().Get<Intrinsic.CNOT>().Controlled;
+            var args = op.__dataIn((controls, (control, target)));
+            var expected = new RuntimeMetadata()
+            {
+                Label = "X",
+                FormattedNonQubitArgs = "",
+                IsAdjoint = false,
+                IsControlled = true,
+                IsMeasurement = false,
+                IsComposite = false,
+                Children = null,
+                Controls = controls.Append(control),
+                Targets = new List<Qubit>() { target },
+            };
+
+            Assert.True(op.GetRuntimeMetadata(args) == expected);
+        }
+
+        [Fact]
+        public void ControlledCCNOT()
+        {
+            Qubit control1 = new FreeQubit(0);
+            Qubit control2 = new FreeQubit(1);
+            Qubit control3 = new FreeQubit(2);
+            Qubit target = new FreeQubit(3);
+            IQArray<Qubit> controls = new QArray<Qubit>(new[] { control1 });
+            var op = new QuantumSimulator().Get<Intrinsic.CCNOT>().Controlled;
+            var args = op.__dataIn((controls, (control2, control3, target)));
+            var expected = new RuntimeMetadata()
+            {
+                Label = "X",
+                FormattedNonQubitArgs = "",
+                IsAdjoint = false,
+                IsControlled = true,
+                IsMeasurement = false,
+                IsComposite = false,
+                Children = null,
+                Controls = new List<Qubit>() { control1, control2, control3 },
+                Targets = new List<Qubit>() { target },
+            };
+
+            Assert.True(op.GetRuntimeMetadata(args) == expected);
+        }
+
+        [Fact]
+        public void ControlledUDT()
+        {
+            IQArray<Qubit> controls = new QArray<Qubit>(new[] { new FreeQubit(0) });
+            Qubit target = new FreeQubit(1);
+            var op = new QuantumSimulator().Get<Tests.Circuits.FooUDTOp>().Controlled;
```

Comment:
although possible in C#, this (a controlled UDT) is not a valid Q# scenario. We should probably delete this test.